File: backend/models/Donor.js
import db from '../database/db.js';

class Donor {
    static async getAll(filters = {}) {
        let query = 'SELECT * FROM donors WHERE 1=1';
        const params = [];

        if (filters.blood_type) {
            query += ' AND blood_type = ?';
            params.push(filters.blood_type);
        }

        if (filters.city) {
            query += ' AND city LIKE ?';
            params.push(`%${filters.city}%`);
        }

        if (filters.available !== undefined) {
            query += ' AND available = ?';
            params.push(filters.available ? 1 : 0);
        }

        query += ' ORDER BY created_at DESC';

        return await db.query(query, params);
    }

    static async getById(id) {
        return await db.get('SELECT * FROM donors WHERE id = ?', [id]);
    }

    static async getByUserId(userId) {
        return await db.get('SELECT * FROM donors WHERE user_id = ?', [userId]);
    }

    static async create(donor) {
        const result = await db.run(`
            INSERT INTO donors (user_id, name, blood_type, phone, email, city, address, available, last_donation)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        `, [
            donor.user_id || null,
            donor.name,
            donor.blood_type,
            donor.phone,
            donor.email || null,
            donor.city,
            donor.address || null,
            donor.available !== undefined ? (donor.available ? 1 : 0) : 1,
            donor.last_donation || null
        ]);

        return { id: result.lastInsertRowid, ...donor };
    }

    static async update(id, donor) {
        const fields = [];
        const params = [];

        if (donor.name) { fields.push('name = ?'); params.push(donor.name); }
        if (donor.blood_type) { fields.push('blood_type = ?'); params.push(donor.blood_type); }
        if (donor.phone) { fields.push('phone = ?'); params.push(donor.phone); }
        if (donor.email !== undefined) { fields.push('email = ?'); params.push(donor.email); }
        if (donor.city) { fields.push('city = ?'); params.push(donor.city); }
        if (donor.address !== undefined) { fields.push('address = ?'); params.push(donor.address); }
        if (donor.available !== undefined) { fields.push('available = ?'); params.push(donor.available ? 1 : 0); }
        if (donor.last_donation !== undefined) { fields.push('last_donation = ?'); params.push(donor.last_donation); }

        if (fields.length === 0) return await this.getById(id);

        params.push(id);
        await db.run(`UPDATE donors SET ${fields.join(', ')} WHERE id = ?`, params);

        return await this.getById(id);
    }

    static async delete(id) {
        return await db.run('DELETE FROM donors WHERE id = ?', [id]);
    }

    static async getByBloodType(bloodType) {
        return await db.query('SELECT * FROM donors WHERE blood_type = ? AND available = 1', [bloodType]);
    }

    static async getStats() {
        const total = await db.get('SELECT COUNT(*) as count FROM donors');
        const available = await db.get('SELECT COUNT(*) as count FROM donors WHERE available = 1');
        const byType = await db.query(`
            SELECT blood_type, COUNT(*) as count 
            FROM donors 
            WHERE available = 1 
            GROUP BY blood_type
        `);

        return {
            total: total.count,
            available: available.count,
            byType: byType.reduce((acc, item) => {
                acc[item.blood_type] = item.count;
                return acc;
            }, {})
        };
    }
}

export default Donor;

File: backend_python/models/donor.py
"""
Donor Model for BEOS Python Backend
"""

from database.db import get_db
from typing import Optional, Dict, Any, List


class Donor:
    """Donor model for blood donors"""
    
    @staticmethod
    async def get_all(filters: Dict[str, Any] = None) -> List[Dict[str, Any]]:
        """Get all donors with optional filters"""
        db = await get_db()
        filters = filters or {}
        
        query = "SELECT * FROM donors WHERE 1=1"
        params = []
        
        if filters.get("blood_type"):
            query += " AND blood_type = ?"
            params.append(filters["blood_type"])
        
        if filters.get("city"):
            query += " AND city LIKE ?"
            params.append(f"%{filters['city']}%")
        
        if filters.get("available") is not None:
            query += " AND available = ?"
            params.append(1 if filters["available"] else 0)
        
        query += " ORDER BY created_at DESC"
        
        cursor = await db.execute(query, params)
        rows = await cursor.fetchall()
        return [dict(row) for row in rows]
    
    @staticmethod
    async def get_by_id(donor_id: int) -> Optional[Dict[str, Any]]:
        """Get donor by ID"""
        db = await get_db()
        cursor = await db.execute("SELECT * FROM donors WHERE id = ?", (donor_id,))
        row = await cursor.fetchone()
        if row:
            return dict(row)
        return None
    
    @staticmethod
    async def get_by_user_id(user_id: int) -> Optional[Dict[str, Any]]:
        """Get donor by user ID"""
        db = await get_db()
        cursor = await db.execute("SELECT * FROM donors WHERE user_id = ?", (user_id,))
        row = await cursor.fetchone()
        if row:
            return dict(row)
        return None
    
    @staticmethod
    async def create(donor: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new donor"""
        db = await get_db()
        
        cursor = await db.execute(
            """INSERT INTO donors (user_id, name, blood_type, phone, email, city, address, available, last_donation)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                donor.get("user_id"),
                donor["name"],
                donor["blood_type"],
                donor["phone"],
                donor.get("email"),
                donor["city"],
                donor.get("address"),
                1 if donor.get("available", True) else 0,
                donor.get("last_donation")
            )
        )
        await db.commit()
        
        return {"id": cursor.lastrowid, **donor}
    
    @staticmethod
    async def update(donor_id: int, donor: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Update a donor"""
        db = await get_db()
        
        fields = []
        params = []
        
        if donor.get("name"):
            fields.append("name = ?")
            params.append(donor["name"])
        if donor.get("blood_type"):
            fields.append("blood_type = ?")
            params.append(donor["blood_type"])
        if donor.get("phone"):
            fields.append("phone = ?")
            params.append(donor["phone"])
        if "email" in donor:
            fields.append("email = ?")
            params.append(donor["email"])
        if donor.get("city"):
            fields.append("city = ?")
            params.append(donor["city"])
        if "address" in donor:
            fields.append("address = ?")
            params.append(donor["address"])
        if "available" in donor:
            fields.append("available = ?")
            params.append(1 if donor["available"] else 0)
        if "last_donation" in donor:
            fields.append("last_donation = ?")
            params.append(donor["last_donation"])
        
        if not fields:
            return await Donor.get_by_id(donor_id)
        
        params.append(donor_id)
        await db.execute(f"UPDATE donors SET {', '.join(fields)} WHERE id = ?", params)
        await db.commit()
        
        return await Donor.get_by_id(donor_id)
    
    @staticmethod
    async def delete(donor_id: int) -> bool:
        """Delete a donor"""
        db = await get_db()
        await db.execute("DELETE FROM donors WHERE id = ?", (donor_id,))
        await db.commit()
        return True
    
    @staticmethod
    async def get_by_blood_type(blood_type: str) -> List[Dict[str, Any]]:
        """Get available donors by blood type"""
        db = await get_db()
        cursor = await db.execute(
            "SELECT * FROM donors WHERE blood_type = ? AND available = 1",
            (blood_type,)
        )
        rows = await cursor.fetchall()
        return [dict(row) for row in rows]
    
    @staticmethod
    async def get_stats() -> Dict[str, Any]:
        """Get donor statistics"""
        db = await get_db()
        
        cursor = await db.execute("SELECT COUNT(*) as count FROM donors")
        total = (await cursor.fetchone())[0]
        
        cursor = await db.execute("SELECT COUNT(*) as count FROM donors WHERE available = 1")
        available = (await cursor.fetchone())[0]
        
        cursor = await db.execute(
            """SELECT blood_type, COUNT(*) as count 
            FROM donors 
            WHERE available = 1 
            GROUP BY blood_type"""
        )
        by_type_rows = await cursor.fetchall()
        by_type = {row[0]: row[1] for row in by_type_rows}
        
        return {
            "total": total,
            "available": available,
            "byType": by_type
        }
